### DISCUSSION TITLE
Database related topic

### DISCUSSION THREAD

**Discussion by wormstein** [EdNekebno/DGTCentaurMods#58] [General]
Hi,
Before I forget how you suggest should we do with the database on install? 

We could either flush all tables on building the release package or store an empty one in git...

It's amazing work you did in your last commit! Makes me feel bad i don't have that much time available :))

**Comment by EdNekebno** [comment]
I thought I commented here...but it doesn't seem to be here.

I think my github newness means I uploaded a version with two new tables in. I intended it to remain the basic one that you created. SQLAlchemy can create the tables from the classes defined in model.py , so leaving them out is the best option as we can make it magically create them.

**Comment by EdNekebno** [comment]
Oh now I'm looking at stockfish I understand some more. We should empty the tables but have the table structure there. Because stockfish can't set them up - otherwise the end user will have to use another script first before it starts to log stockfish.

I also see what you were saying about takebacks, etc, centaur sends a move list each time. I'll make stockfish break the movelist back out to moves and update the tables. It just loses the exact time of the move, but that's not too bad.

**Comment by wormstein** [comment]
Yeah, flushing I meant just wipe the data, not the structure. Or if you have the sqk scruot to build the database, I could call it after the .deb packahe is deployed. Or maybe a python file to use sqlalchemy.

**Comment by EdNekebno** [comment]
That's voodoo to me :) It should be easy enough to generate a python script that empties the tables though.

**Comment by wormstein** [comment]
Git will probably cry that it will overwrite your db file. I think to solve this you have to look at the following commands depending on situarion to resolve any conflicts.
`
git checkout --theirs <path_to_centaur.db>
git checkout --ours <path_to_centaur,db>
`
Or have an empty DB in here and one that you just replace it in your local git repo before commit and push.
It's a tricky situation though.

I would like this:
git rm centaur.db, commit and push so it ggets removed.
add it to ignore list
force commit an empty database (git commit -f) then push

Now, everytime you do a change in db structure, you can force commit a new one with git add -f centaur.db

After we have a starting structure of the database, a very good practice is having a SQL update script (create, remove tables, move data as necessary, etc) that can be run with any updates to software. This way we can preserve any data that a user might already have in the DB when we release an update. Is called database migration/update. This might be something we need to considerr.

**Comment by EdNekebno** [comment]
Well I appear to have done it in a roundabout way with a million pushes. But I believe the idea now is that every time the structure changes we just create one with the structure and force commit it?

**Reply by wormstein** [reply]
Yes, because the database is in the ignore list now you just have to force it when staging, when needed,

`
git add -f centaur.db
git commit
git push
`

**Comment by wormstein** [comment]
Performance concerns.
I have some concerns storing all moves in a database structure though. RPi is not a very powerful CPU and quite limited memorywise to handle such a big table. It might work fine in the beggining but thinking that a game has an average of 120 moves, just 50 games will create over 5000 rows. Doing a select on that one.... makes me thinking.

That's why I was wondering yesterday why not rely on pgn files (and games in db), eventually stored in folders by date so we don't overrun the filesystem.

**Comment by EdNekebno** [comment]
I had a similar thought originally. Then in my mind was the option that we can purge the oldest stuff if we need to.
However, 5000 rows, for example, sounds big but it's tiny for a database. Sqlite is capable of much much much much more. In terms of how we access those particular tables, it's also one of the simplest possible access (get all data from this table for this game). Then as you indicate the RPi would be the limiting factor. But i'm not sure it's that much of one. When we read the data back we don't need super fast speeds - 1 second or so is quite reasonable.

It seems to me that this is testable. I'll make a script to run some test data and see what happens.

**Reply by wormstein** [reply]
Sounds like a plan. We could work later on a settings web page where the user can select the number of games to save...

**Reply by EdNekebno** [reply]
I pumped it full of some test data. Interestingly 21 games gave me a time to retrieve the moves for a game as 0.07 seconds. 1021 games gave 0.06 seconds. Then I think as long as it fits in memory then indexing has our back because of the very simplistic way we need to read the data - we get all moves for a particular game. My 1021 game database is 13MB . I think that 120 moves is actually on the high side for an average game, but I'd say we can arbitrarily cap it at 10000 games. The db would handle it fairly easily, but it's such a big figure that nearly nobody would hit it and for those that do that data really doesn't have any relevance anymore.

**Reply by wormstein** [reply]
Sorry about that... you're right. If it's average, that is lower... not 120.
Very good times!
